tests/test_downloadermiddleware.py

```diff
@@ -3,33 +3,33 @@
 
 from pytest import mark
 from twisted.internet import defer
-from twisted.internet.defer import Deferred
+from twisted.internet.defer import Deferred, inlineCallbacks
 from twisted.python.failure import Failure
 from twisted.trial.unittest import TestCase
 
 from scrapy.core.downloader.middleware import DownloaderMiddlewareManager
 from scrapy.exceptions import _InvalidOutput
 from scrapy.http import Request, Response
-from scrapy.spiders import Spider
 from scrapy.utils.python import to_bytes
 from scrapy.utils.test import get_crawler, get_from_asyncio_queue
+from tests.spiders import NoRequestsSpider
 
 
 class ManagerTestCase(TestCase):
     settings_dict = None
 
+    @inlineCallbacks
     def setUp(self):
-        self.crawler = get_crawler(Spider, self.settings_dict)
-        self.spider = self.crawler._create_spider("foo")
+        self.crawler = get_crawler(NoRequestsSpider, self.settings_dict)
+        yield self.crawler.crawl()
+        self.spider = self.crawler.spider
         self.mwman = DownloaderMiddlewareManager.from_crawler(self.crawler)
-        # some mw depends on stats collector
-        self.crawler.stats.open_spider(self.spider)
-        return self.mwman.open_spider(self.spider)
+        yield self.mwman.open_spider(self.spider)
 
     def tearDown(self):
-        self.crawler.stats.close_spider(self.spider, "")
         return self.mwman.close_spider(self.spider)
```

Comment:
Should we close the crawler itself? 
A related question: instead of creating DownloaderMiddlewareManager, should we get an instance which is created by the crawler?

Reply:
> Should we close the crawler itself?

The crawler should be "closed" when `yield crawl()` returns.

> instead of creating DownloaderMiddlewareManager, should we get an instance which is created by the crawler?

That instance is created by the engine, so we can do it if we decide to keep calling `crawler.crawl()` (which I was going to revert and just call `_load_settings()`).

Reply:
> The crawler should be "closed" when yield crawl() returns.

oh, right; I think most of my comments to the PR are irrelevant :)

Reply:
> The crawler should be "closed" when yield crawl() returns.

Hm, why would it be guaranteed? It seems crawler.crawl waits until `self.engine.start` is successful, but it doesn't wait for the crawling to be stopped.

Reply:
`ExecutionEngine.start()` returns a deferred that fires when the crawling finishes.

Reply:
Overall, it seems we're doing some strange things in the tests, which is not very clean: 

* crawler is started and stopped(?)
* then some internal components are created using this stopped crawler (which we don't have a well-defined behavior for)

It seems we're not using our public APIs (or any well-defined APIs), and instead just trying to get the components in usable state, using whatever the current inernal behavior is.

I haven't understood the recent proposals about `_load_settings`, maybe it makes sense to try them.

Another alternative would be to try to use public APIs. If it's not possible with the current .crawl API, it could be an argument to improve Crawler API slightly, e.g. split .crawl() into a few methods, e.g. to init the crawler and actually start it (.crawl sould be kept as a shortcut to do both).

Reply:
> Overall, it seems we're doing some strange things in the tests, which is not very clean:

Yes, this is my main problem with the current PR code and the main reason I want a separate initialization function.

> Another alternative would be to try to use public APIs. If it's not possible with the current .crawl API, it could be an argument to improve Crawler API slightly, e.g. split .crawl() into a few methods, e.g. to init the crawler and actually start it (.crawl sould be kept as a shortcut to do both).

That's what I proposed initially in https://github.com/scrapy/scrapy/pull/6038#issuecomment-1710108413 and recently in https://github.com/scrapy/scrapy/pull/6038#issuecomment-1713601583

Though I see it as a private function that includes all that initialization we are moving from `__init__`, not something we expect the user to call. If we want public functions, I think it's better to reshuffle the code so that we have e.g. only `__init__()` and `crawl()`, not 3 functions, but I'm not sure how useful would that be.

Reply:
To be more concrete, I was thinking about doing it like this:

```py
    def init_crawl(self):  # we can also make it private
        # + exception and flag handling
        self.spider = self._create_spider(*args, **kwargs)
        self.engine = self._create_engine()
        # all the new initialization code moved from __init__ goes here?

    @defer.inlineCallbacks
    def start_engine(self):  # we can also make it private
        # + exception and flag handling
        start_requests = iter(self.spider.start_requests())
        yield self.engine.open_spider(self.spider, start_requests)
        yield defer.maybeDeferred(self.engine.start)

    @defer.inlineCallbacks
    def crawl(self):
        self.init_crawl()
        yield self.start_engine()
```

Is the idea to just move init_crawl content to `__init__`? On a first sight, it looks more risky, as it'd start calling `spider.__init__` and `spider.from_crawler`, so we'd need to check if that's backwards compatible, and what kind of issues it can raise.

Reply:
2 separate private functions look better to me than 1 private function (_init_crawl + keep engine starting code in `crawl()`), because this way we'd 

1) keep the code in crawl() on the same level of abstraction (no low-level vs high-level code)
2) allow to do the init, then do something else, and then start the crawl (though not sure what's exactly the use case)

Reply:
The current idea is

```python
    def _load_settings(self) -> None:
        if self._settings_loaded:
            return
        self._settings_loaded = True
        # all the new initialization code moved from __init__ goes here

    def crawl(self, *args: Any, **kwargs: Any) -> Generator[Deferred, Any, None]:
        # flag handling
        try:
            self.spider = self._create_spider(*args, **kwargs)
            self._load_settings()
            self.engine = self._create_engine()
            # etc.
```

so the `Crawler` API is the same but `scrapy shell` and `scrapy.utils.test.get_crawler()` can call `crawler._load_settings()` (and `crawler.spider = crawler._create_spider()` if needed) to get an initialized Crawler instance. This doesn't create or start the engine, but the old API didn't either.

We could move `_create_spider` and `_create_engine` there too, to slightly simplify the code in `scrapy shell` and to allow tests to get more components from the engine, though this state may still be not suitable for some tests (e.g. `engine.scheduler` is only assigned in `engine.open_spider()`) and also making such changes to tests is out of scope.

> Is the idea to just move init_crawl content to __init__

I don't want to do this but I proposed it as a way to have no more than 2 public methods.

Reply:
This sounds fine. 
Should it be called something like _apply_settings though?

Reply:
No opinion on the name (I proposed load, but apply sounds good too).

Reply:
@kmike do you think it should create the spider and the engine instances?

Reply:
@wRAR no strong opinion on that; please follow the approach you prefer. I haven't checked all code which is relevant (e.g. scrapy shell), while you have checked it :)